custom-checks/checkstyle/src/main/java/org/openhab/tools/analysis/checkstyle/ThingHandlerSetLabelCheck.java

```diff
@@ -0,0 +1,212 @@
+
+                if (currentFileClassName.equals(baseClassName) && areFilesFromSamePackage) {
+                    DetailAST fileRootAST = files.get(file);
+                    classesExtendingBaseThingHandler.put(file, fileRootAST);
+                    visitedFiles.add(file);
+                }
+            });
+        }
+    }
+
+    private void findDerivedAndBaseClasses() {
+        files.forEach((file, rootAST) -> {
+            Optional<String> fullyQualifiedName = getFullyQualifiedName(rootAST);
+            Optional<String> baseClassName = getBaseClassName(rootAST);
+            String currentFullName = fullyQualifiedName.orElse(UNKNOWN);
+            String currentBaseName = baseClassName.orElse(UNKNOWN);
+            // if the currentFullName is not UNKNOWN our SEARCHED PACKAGE is found for the current class
+            boolean isHandlerImportsBaseThingHanlderPackage = !UNKNOWN.equals(currentFullName);
+            // check for case where the SEARCHED PACKAGE is imported but not extended
+            boolean isHandlerExtendingBaseThingHandler = !UNKNOWN.equals(currentBaseName)
+                    && BASE_THING_HANDLER.equals(currentBaseName);
+
+            if (isHandlerImportsBaseThingHanlderPackage && isHandlerExtendingBaseThingHandler) {
+                derivedAndBaseClasses.put(file, currentFullName);
+            } else {
+                derivedAndBaseClasses.put(file, currentBaseName);
+            }
+        });
+    }
+
+    private Optional<String> getFullyQualifiedName(DetailAST rootAST) {
+        Queue<DetailAST> visitedNodes = new LinkedList<>();
+        visitedNodes.add(rootAST);
+
+        while (!visitedNodes.isEmpty()) {
+            DetailAST currentNode = visitedNodes.poll();
+            String currentFullName = CheckUtils.createFullType(currentNode).getText();
+
+            if (SEARCHED_PACKAGE.equals(currentFullName)) {
+                return Optional.of(SEARCHED_PACKAGE);
+            } else {
+                if (currentNode.getNextSibling() != null) {
+                    visitedNodes.add(currentNode.getNextSibling());
+                }
+            }
+        }
+        return Optional.empty();
+    }
+
+    private Optional<String> getBaseClassName(DetailAST rootAST) {
+        Queue<DetailAST> visitedNodes = new LinkedList<>();
+        visitedNodes.add(rootAST);
+
+        while (!visitedNodes.isEmpty()) {
+            DetailAST currentNode = visitedNodes.poll();
+
+            if (currentNode.getType() == TokenTypes.CLASS_DEF) {
+                DetailAST extendsClause = currentNode.findFirstToken(TokenTypes.EXTENDS_CLAUSE);
+
+                if (extendsClause != null) {
+                    String inheritedClassName = extendsClause.getFirstChild().getText();
+                    // Only the simple name of the class can be retrieved from the extends clause
+                    return Optional.of(inheritedClassName);
+                }
```

Comment:
Should you maybe return here as well? Or should we check for all classes in a single file what if we we put the parent and child class in a single file and one extra to make it hard, will it work?

```java
class Test extends Object {
}

public class SomeThingHandler extends TestThingHandler { 
   public SomeThingHandler(){
      thing.setLabel("test");
   }
}

public class TestThingHandler extends BaseThingHandler {
}

Reply:
No, it will not work because I am currently searching for the first extend.